src/main/java/bio/terra/cda/app/service/QueryService.java

```diff
@@ -54,24 +72,38 @@ private Job runJob(Job queryJob) {
     }
   }
 
-  public List<String> runQuery(Query query) {
-    String queryString = generateQueryClause(query);
+  private Integer defaultValue(Integer val, Integer def) {
+    if (val != null) {
+      return val;
+    } else {
+      return def;
+    }
+  }
+
+  public QueryResult runQuery(Query query, Integer offset, Integer limit) {
+
+
+    String queryString = (new QueryTranslator(this.table, query)).sql();
+
     // Wrap query so it returns JSON
-    String jsonQuery = String.format("SELECT TO_JSON_STRING(t,true) from (%s) as t", queryString);
+    offset = defaultValue(offset, 0);
+    limit = defaultValue(limit, 100);
+    String jsonQuery =
+        String.format(
+            "SELECT TO_JSON_STRING(t,true) from (%s LIMIT %s OFFSET %s) as t",
+            queryString, limit, offset);
     QueryJobConfiguration queryConfig =
-        QueryJobConfiguration.newBuilder(jsonQuery).setUseLegacySql(true).build();
+        QueryJobConfiguration.newBuilder(jsonQuery).setUseLegacySql(false).build();
 
     // Create a job ID so that we can safely retry.
     JobId jobId = JobId.of(UUID.randomUUID().toString());
-    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
-
-    queryJob = runJob(queryJob);
 
-    return getJobResults(queryJob);
-  }
-
-  private String generateQueryClause(Query query) {
-    // FIXME need to implement
-    return "SELECT X from Y WHERE Z";
+    try {
+      Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+      queryJob = runJob(queryJob);
+      return new QueryResult(jsonQuery, getJobResults(queryJob));
```

Comment:
I think this change got lost in the merge. I was thinking that we don't want to return the query string that contains the `TO_JSON_STRING` wrapper, since that isn't part of the user's query. So we should return `queryString` here and below in the thrown error. Although, to be accurate `queryString` should contain the OFFSET and LIMIT parts of the query too.